### PR TITLE
Add sextOrBitCast to conversions of Builtin.Word

### PR DESCRIPTION
**Pull request by Kaiede**
An error was hit when attempting to build Swift on a 32-bit Linux host. It was asserting when attempting to run ‘RedunantLoadElim” during the optimizer. The reason why is a bit messy.

First, Builtin.Word is always considered to be 64-bit when it is a SILType. This means the optimizer, when working with a Builtin.Word, will create 64-bit APInts when converting from other types.

Second, the SIL being output for a particular literal looked like this:
Int2048 (integer_literal) : Int32 (Signed Truncation) : Word (zextOrBitCast)

The constant fold behavior would convert this into an integer_literal of the Builtin.Word type. But because of the zext cast, if the original literal was -1, it would become +4 billion during folding.

This normally isn’t a problem **except** when SIL is still attempting to optimize. (See First)

Third, The Redundant Load Elimination pass was attempting to make sense of an index_addr instruction which was indexing at -1. This happens when you perform “-= 1” on an UnsafePointer. Because SIL was interpreting Word’s size incorrectly, it caused the RLE pass to ask “what is at ptr[4billion]?” Since there’s no feasible answer for such a question, the answer was “nothing”, which tripped the assert.

This fix uses sign extension when converting “upward” and the Integer type is signed, otherwise it will still use zero extension.

Resolves [SR-8870](https://bugs.swift.org/browse/SR-8870).

### CODE REVIEW

**Comment by Kaiede** [comment]
FYI, I'm still doing a couple last checks on this change. It should be fine, but I re-enabled the zext conversions after my last build.

File: stdlib/public/core/IntegerTypes.swift.gyb [hunk 1]
```diff
@@ -1616,6 +1616,8 @@ ${assignmentOperatorComment(x.operator, True)}
   init(_ _v: Builtin.Word) {
 % if BuiltinName == 'Int32':
     self._value = Builtin.truncOrBitCast_Word_Int32(_v)
+% elif BuiltinName == 'Int64' and signed:
+    self._value = Builtin.sextOrBitCast_Word_Int64(_v)
 % elif BuiltinName == 'Int64':
     self._value = Builtin.zextOrBitCast_Word_Int64(_v)
```
[the remark this answers is not in the document]
Comment (Kaiede) [reply stdlib/public/core/IntegerTypes.swift.gyb]: Not sure how I missed that. That is more correct, yes. Thanks.

**Comment by milseman** [comment]
A test may have to be written in SIL for the iPhone simulator, if that's the only place this change is visible on Darwin.

@eeckstein who can advise?

**Review by jckarter — APPROVED** [review]
LGTM

**Comment by Kaiede** [comment]
I've updated with the feedback from @xwu and my last set of checks with that version are good.

@milseman Regarding the test, I'm not 100% sure that writing it in SIL really helps? The change only affects the SIL generated from the Int32/UInt32 types, and so any test in SIL is really testing the "Builtin.Word is always 64-bit during Optimization" behavior during constant folding. Although maybe @eeckstein has a better idea how to write a test to prevent regressions here.

**Comment by jckarter** [comment]
I feel like a source-level test would be the most robust thing. Execution-testing some Swift code that exercises one of the formerly problematic situations should be more robust and ensure that any refactorings we do in the future here don't break things even if all the implementation details completely change.

**Comment by Kaiede** [comment]
So effectively a snippet where if it compiles, the test passes, and if it doesn't compile, the test fails? Something like the following maybe?

```
func uninlinableFunc(ptr: UnsafePointer<UInt8>) {
    ptr -= 1 // Causes an index_addr of -1 to be emitted.
    print(ptr.pointee) // Causes us to load the result of index_addr after optimization
}

let base: UnsafePointer<UInt8> = // Allocate some memory
// Write Something Interesting into the elements of memory
let index = base + 2
uninlinableFunc(ptr: index)
```

**Comment by Kaiede** [comment]
> This fix is in the stdlib.
> But we should also fix the crash in RLE. Optimizations should not crash for any swift input code.

Which is fair, although the fix for SIL Optimizer is a bit more complicated (the thread is gnarly with details). I kinda hope the plan there wouldn't be to just abort the optimization if given a nonsense location to check for loading? The real fix should be to actually fix it so that Builtin.Word is the correct size for the target so constant folding doesn't produce the nonsense, but seeing as that's been marked with FIXMEs already, I suspect someone's already looked at the amount of work *that* is and decided against it at that time.

My thoughts here are that the RLE piece is behaving as expected. What do you think the fix is here?

**Comment by Kaiede** [comment]
Turns out there’s already a test for RLE which catches this issue, it’s just a bit hard to run when stdlib is failing to build. So the SIL test for this does exist, and does fail, even with the stdlib fix. So, I’ll eat my hat a bit, since I guess it’s also established in the code history that this optimization shouldn’t be failing like this.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by Kaiede** [comment]
Ugh, it looks like the CI doesn't like that I did an amend commit here. Lesson Learned. ☹️

**Comment by jckarter** [comment]
No problem, it restarts automatically.

**Comment by Kaiede** [comment]
Yeah, just weird that it was trying to sync a commit that no longer exists.

**Comment by jckarter** [comment]
Yeah, that's unfortunately a bug in the Jenkins git plugin we've had to live with for a while.

**Pull request merged**